### PR TITLE
feat: Add client-side Yosys WASM synthesis for Tauri desktop builds

### LINKED ISSUE #1021: feat: Client-side Verilog synthesis via Yosys WASM for Tauri desktop builds
### Is your feature request related to a problem? Please describe

In the Tauri desktop build, the Verilog synthesis feature fails silently because it depends on the Rails server endpoint `/api/v1/simulator/verilogcv` which is unreachable when the app is running offline or without a backend. Desktop users cannot compile any Verilog code.

### Describe the solution you'd like

Add a client-side synthesis path using @yowasp/yosys — a WebAssembly port of Yosys that runs entirely inside the browser/desktop app without any server dependency.
-> Architecture:
- Tauri Desktop: Verilog → Web Worker → Yosys WASM → yosys2digitaljs → CV render
- Web Browser: Verilog → Rails API → Yosys server → CV render (unchanged)
-> The routing is automatic based on environment detection via window.__TAURI__.

### Describe alternatives you've considered

Bundling a native Yosys binary with Tauri was considered but rejected due to platform-specific complexity. The WASM approach works cross-platform with no native dependencies.

### Is requested feature related to vue simulator?

Yes — specifically the Verilog editor and synthesis pipeline in the Vue simulator, affecting the Tauri desktop build.

### Additional context

_No response_

### Are you working on this?

Yes

### PR DESCRIPTION
**Pull request by SouparnaChatterjee**
Fixes #1021 

 

## Summary
Implements client-side Verilog synthesis using Yosys WebAssembly (@yowasp/yosys) for the Tauri desktop build, eliminating the dependency on the Rails server for Verilog compilation.  This PR adds a fully client-side synthesis path using Yosys WebAssembly, so the desktop app can compile Verilog with zero server dependency.

### `public/yosys-worker.js` (new file)
- Web Worker that loads @yowasp/yosys WASM in a separate thread
- Receives Verilog code, runs full synthesis pipeline
- Queues messages until WASM is loaded
- Uses `hierarchy -top` or `-auto-top` for correct multi-module handling
- Returns synthesized JSON without blocking the main UI thread
- Sends `ready` message when WASM is loaded

### `src/simulator/src/Verilog2CV.js`
- >`shouldUseWasm()` — detects Tauri via `window.__TAURI__`
- >`initWasmWorker()` — single shared Promise, prevents duplicate polling
- >`synthesizeWithWasm()` — full client-side synthesis path
-> `synthesizeWithServer()` — original server path, completely unchanged
- >`extractTopModule()` — infers top module, falls back to `-auto-top`
- >`convertYosysToDigitalJs()` — converts raw Yosys JSON to CV format, handles constant bits ("0","1","x","z") as Constant devices
- >`computeLayeredLayout()` — topological sort of netlist, assigns gates to layers based on signal flow (inputs → logic → outputs), computes pixel coordinates before CV elements are instantiated
- >`renderVerilogCircuit()` — shared rendering for both paths

### `scripts/download-yosys-wasm.js`
-> Postinstall script copying WASM files from `@yowasp/yosys` into `public/`
- >Skips gracefully on Netlify/web builds (Tauri-only feature)
- >WASM files are gitignored to avoid committing ~50MB of binaries

## Architecture
Tauri Desktop(*new):
Verilog -> yosys-worker.js (WASM) ->convertYosysToDigitalJS-> computeLayeredLayout-> YosysJSON2CV ->CV canvas

Web Browser (unchanged):
Verilog-> Rails API /verilogcv ->Yosys server ->YosysJSON2CV -> CV canvas

### Screenshots of the UI changes (If any) -

Before:
Screenshot 2026-03-17 211521

After:(full adder synthesized entirely client-side, zero server call)
Screenshot 2026-03-17 211323

## Testing
->AND gate, half adder, full adder tested on Tauri desktop
->Verified offline — WiFi disabled, synthesis still completes
->Zero network calls to `/api/v1/simulator/verilogcv` during WASM path
->Server fallback works correctly in browser when `window.__TAURI__` is absent
->Netlify web build unaffected (postinstall skips gracefully)
->`window.__yosysWasmEnabled = true` override for dev testing without Tauri

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [x] No, I wrote all the code myself
- [] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [] I have reviewed every single line of the AI-generated code
- [] I can explain the purpose and logic of each function/component I added
- [] I have tested edge cases and understand how the code handles them
- [] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**

The problem: Tauri desktop app cannot reach the Rails server for Verilog synthesis, so the feature silently fails offline.

I chose a Web Worker architecture because Yosys synthesis is CPU-intensive and would block the UI thread if run directly. The worker runs in the background and communicates results via postMessage.

The key design decision was keeping both paths (WASM and server) working simultaneously, with automatic routing based on environment detection. This means the web version behavior is completely unchanged.

I used `window.__TAURI__` for environment detection since that global is injected by the Tauri runtime — the standard way to detect desktop context in this codebase.

The `shouldUseWasm()` function also supports `window.__yosysWasmEnabled` as a development override flag for testing without a full Tauri build.
---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [ ] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

---

Note: Please check **Allow edits from maintainers** if you would like us to assist in the PR.

## Summary by CodeRabbit

* **New Features**
  * In-browser WASM synthesis via a background worker; client-side Verilog → circuit conversion and optional local linting/processing utilities; installer script to fetch WASM assets.

* **Bug Fixes**
  * Better validation and error reporting for synthesis and conversion; missing ports/connectors produce warnings; more resilient rendering and layout centering.

* **Chores**
  * Dependency and script updates; added ignore rules for local WASM artifacts.

### CODE REVIEW

File: scripts/download-yosys-wasm.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  var src  = path.join(srcDir, file.src);
+  var dest = path.join(destDir, file.dest);
+  if (fs.existsSync(src)) {
+    fs.copyFileSync(src, dest);
+    var size = (fs.statSync(dest).size / 1024 / 1024).toFixed(1);
+    console.log('Copied ' + file.dest + ' (' + size + ' MB)');
+  } else {
+    console.warn('WARNING: ' + file.src + ' not found at ' + src);
+  }
+});
```
[the remark this answers is not in the document]
Comment (SouparnaChatterjee) [reply scripts/download-yosys-wasm.js:34]: The two failure modes are intentionally handled differently:-
1) @yowasp/yosys not installed (e.g. Netlify/web builds) → exit(0) — this is expected, WASM synthesis is a Tauri-only feature and web builds should not fail because of it.
2) @yowasp/yosys is installed but a file is missing → this is already caught by the console.warn and the copy simply skips. Since this only runs in Tauri desktop builds where the package IS installed, a missing file would be caught at runtime when the worker tries to load /yosys-bundle.js.
-> The web build path (exit(0)) is the correct behavior here — Netlify should not fail because a Tauri-only dependency is absent.

File: public/yosys2digitaljs.js [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+exports.verilator_lint = verilator_lint;
+exports.process = process;
+exports.process_files = process_files;
+exports.process_sv = process_sv;
+const tmp = require("tmp-promise");
+const child_process = require("child_process");
+const fs = require("fs");
+const path = require("path");
+const util_1 = require("util");
+const core_1 = require("./core");
```
[the remark this answers is not in the document]
Comment (SouparnaChatterjee) [reply public/yosys2digitaljs.js]: These files (public/yosys2digitaljs.js and public/yosys2digitaljs-core.js) were deleted in commit 91fda7b — they were unused Node.js files incompatible with the browser environment. This issue is no longer relevant.

File: public/yosys2digitaljs.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+const sanitize = require('sanitize-filename');
```
[the remark this answers is not in the document]
Comment (SouparnaChatterjee) [reply public/yosys2digitaljs.js]: These files were deleted in commit 91fda7b. tmp-promise and sanitize-filename are Node.js server-side dependencies that have no place in a browser build. The WASM path uses @yowasp/yosys directly in the worker instead.

**Review by naman79820 — COMMENTED** [review]
Putting this on hold as it's most likely POC for project-7

**Comment by naman79820** [comment]
closing as part of gsoc project :))
